File: slidingWindow/findLongestSubstring.js
// Write a function called findLongestSubstring, which accepts a string and returns the length of the longest substring with all distinct characters.

// Julie's version:
function findLongestSubstring(s){
  let longest = 0;
  let start = 0;
  let seen = {};

  for (let i = 0; i < s.length; i++) {
    const windowSize = i - start + 1;
    const char = s[i];
    if (seen[char] === undefined || seen[char] < start) {
      longest = windowSize > longest ? windowSize : longest;
    } else {
      start = seen[char] + 1;
    }
    seen[char] = i;
  }
  return longest;
}

console.log(findLongestSubstring('')); // 0
console.log(findLongestSubstring('rithmschool')); // 7
console.log(findLongestSubstring('thisisawesome')); // 6
console.log(findLongestSubstring('thecatinthehat')); // 7
console.log(findLongestSubstring('bbbbbb')); // 1
console.log(findLongestSubstring('longestsubstring')); // 8
console.log(findLongestSubstring('thisishowwedoit')); // 6

File: neetCode/slidingWindow/findLongestSubstring.py
def findLongestSubstring(string):
    start = 0
    longest = 0
    seen = {}
    for i in range(len(string)):
        char = string[i]
        if char in seen:
            start = max(start, seen[char] + 1) # essential to have max here otherwise start could go back to the first index of the character, which would make longest too long
        longest = max(longest, (i - start + 1))
        seen[char] = i
    return longest

print(findLongestSubstring('rithmschool')) # 7
print(findLongestSubstring('thisisawesome')) # 6
print(findLongestSubstring('thecatinthehat')) # 7
print(findLongestSubstring('bbbbbb')) # 1
print(findLongestSubstring('longestsubstring')) # 8
print(findLongestSubstring('thisishowwedoit')) # 6